### PR TITLE
Opt-In FireFox Features

### LINKED ISSUE #85: [FireFox] - Opt-In all features
According to the add-on policies, we need all the features provided by the add-on, to be opt-in and not "on by default"

> The features must be “opt-in”, meaning the user has to take non-default action to enact the change. Changes prompted by Firefox after the add-on is installed do not require an additional opt-in. The permissions prompt shown when installing an add-on does not alleviate the need for an opt-in.

### PR DESCRIPTION
**Pull request by Unthrottled**
#### Description

- Changed all settings to be opt-in to conform to the Mozilla Add-On policy.
- Added documentation of the rest of the FireFox features.
- Added a link to said documentation in the settings menu.
- Added a link to show the Options Page in the popup.

#### Motivation and Context

Closes #85 

#### Screenshots (if appropriate):

image

#### Types of changes

- [X] Bug fix (non-breaking change which fixes an issue)
- [X] Breaking change (fix or feature that would cause existing functionality to change)

#### Checklist:

- [X] I updated the changelog with the new functionality.

### CODE REVIEW

File: firefoxThemes/options/options.html
Comment (Unthrottled) [inline firefoxThemes/options/options.html:12]: I chose to link to the `Feature Preview` anchor because, when I choose a lower anchor, all of the images load and the page no-longer is at the specified anchor (real annoying!)

File: firefoxThemes/options/options.js
Comment (Unthrottled) [inline firefoxThemes/options/options.js:33]: I wasn't sure if the themes would re-hydrate from local storage, such that they would have the new key `editorAccentColor`, if it is true that when the new plugin loads, the new definitions get picked up I can remove the fall back.

File: firefoxThemes/popup/index.html
Comment (Unthrottled) [inline firefoxThemes/popup/index.html:41]: Wasn't sure what element to use, so I stuck with span and made it look like  link 😅

Comment (ZimCodes) [reply firefoxThemes/popup/index.html:41]: Ehh, it works.

To make it a bit better though, there should be some kind of affordance here to make it clear it's a clickable item. For instance, adding some hover effects or something. 

Honestly, I prefer the Heart interaction over the link addition. In any case, so long as the Heart navigation is still there I'm fine with this 👍

**Review by ZimCodes — CHANGES_REQUESTED** [review]
Interesting! Did not know there was a policy change in the mix.

For Firefox, there's already a *PERMISSIONS.md* file which explains why each permission in the manifest is necessary. However, the `search` permission in the manifest can be placed as an `optional_permission` instead of a required one. 

There are still some mathematics *(additions & subtractions)* that needs to be performed before approval.

File: firefoxThemes/manifest.json
Comment (ZimCodes) [inline firefoxThemes/manifest.json:4]: Although updated here, the `package.json` file also needs a version update too.

**Review by ZimCodes — CHANGES_REQUESTED** [review]

File: firefoxThemes/manifest.json
Comment (Unthrottled) [reply firefoxThemes/manifest.json:4]: Yeah, I didn't do that because that updates all the things in all the places, I might need to re-think how versioning is handled for a repo that contains distinct products. Might be that FireFox gets its own repository, I dunno yet.

File: firefoxThemes/popup/index.html
Comment (Unthrottled) [reply firefoxThemes/popup/index.html:41]: > Honestly, I prefer the Heart interaction over the link addition

This is just me, but I didn't know to click it until you said something about it, when I see a link, I click and expect to go places. Perhaps because there maybe wasn't anything, other than the cursor changing to button click mode, that this might take me someplace.

File: firefoxThemes/options/options.js
Comment (Unthrottled) [reply firefoxThemes/options/options.js:33]: This comment https://github.com/doki-theme/doki-theme-web/pull/83#discussion_r772809939

Gave me the impression that the `waifuThemes` are persisted in local storage, If I make modifications to existing theme definitions (adding a color), will the existing stored definitions pick up the new changes? Are theme definitions being cached?

File: firefoxThemes/options/options.html
Comment (Unthrottled) [reply firefoxThemes/options/options.html:12]: I'm on the fence about putting documentation into a tooltip that only shows up when hovering over the option. There isn't anything indicating to me that I should hover over something to get more information, other than "Oh I want to try this feature".  But that is me.

> and add a grand ToC somewhere at the beginning of the README file.

Yeah, that sounds like a good idea.

File: firefoxThemes/options/options.js
Comment (ZimCodes) [reply firefoxThemes/options/options.js:33]: > If I make modifications to existing theme definitions (adding a color), will the existing stored definitions pick up the new changes?

Yes
> Gave me the impression that the waifuThemes are persisted in local storage

It is, but every time the browser closes and reopens, dokithemedefinitions gets stored in the local storage again replacing the old one. See ‘startStorage()‘ & ‘WaifuTheme class‘ in resources.js for more info on this.

File: firefoxThemes/manifest.json
Comment (ZimCodes) [reply firefoxThemes/manifest.json:4]: I agree. The Firefox component should have its own separate repo considering how dynamic the theming system is compared to the chromiums. Also considering the fact that majority of recent pull requests are focused around Firefox as well.

File: firefoxThemes/options/options.js
Comment (Unthrottled) [reply firefoxThemes/options/options.js:33]: > It is, but every time the browser closes and reopens

Okay, I am fairly sure that the browser doesn't have to restart to update a plugin, so this is only necessary until the user opens and closes the browser, right?

File: firefoxThemes/resources.js
Comment (Unthrottled) [reply firefoxThemes/resources.js:76]: Got it, I guess I didn't notice that because the checked class gets removed after initialization.

File: firefoxThemes/manifest.json
Comment (Unthrottled) [reply firefoxThemes/manifest.json:4]: Yeah I'll do that later once I prepare another theme release.

**Comment by Unthrottled** [comment]
>  PERMISSIONS.md file which explains why each permission in the manifest is necessary

Does that file do anything or is that just for anybody reviewing the product?

I also moved search to an optional permission, is there any other things that should be optional?

File: firefoxThemes/options/options.js
Comment (ZimCodes) [reply firefoxThemes/options/options.js:33]: > the browser doesn't have to restart to update a plugin

Did not write that.
> this is only necessary until the user opens and closes the browser, right?

This is what I wrote.

**Comment by ZimCodes** [comment]
> Does that file do anything or is that just for anybody reviewing the product?

The latter
> is there any other things that should be optional?

The `browserSettings` permission can also be optional, However it might be a bit tricky(?.) Out of all the *Theme Settings* options, `all`, `light`, & `dark` does not rely on *system* settings instead they use the *browser* settings *(which is the default if the browserSettings permission was not at all added to this extension in the first place.)*. 

So it would be ideal to place the activate this permission only if `druther` or `device` was chosen.

File: firefoxThemes/options/options.js
Comment (Unthrottled) [reply firefoxThemes/options/options.js:33]: I am confused. 

These are the facts that I think I am working with:

- dokithemedefinitions gets stored in the local storage.
-  When browser closes and reopens, dokithemedefinitions gets stored in the local storage again replacing the old one.

The point that I was trying to make is that for existing users of the plugin:

When the the plugin updates, if the browser doesn't restart, then the old dokithemedefinitions in local storage will have `editorAccentColor`.

**Comment by Unthrottled** [comment]
I added browserSetting as an optional permission to be requested only when the user selects the `druthers` or `device` option. I also learned the hard way, that you can only request permissions in a eventHandler.

File: firefoxThemes/options/options.js
Comment (ZimCodes) [reply firefoxThemes/options/options.js:33]: > When the the plugin updates, if the browser doesn't restart, then the old *dokithemedefinitions* in local storage will not have `editorAccentColor`.

Is the `editorAccentColor` property part of the new plugin update? Or is it apart of the old *dokithemedefinitions* one?

**Comment by ZimCodes** [comment]
> I also learned the hard way, that you can only request permissions in a eventHandler.

You've just experience the beauty of an event-driven programming language.

Since it is a small code change, there was an issue when the user selects an option in the *General* category. The *Options Page* would refresh, which, in turn, navigate the user away from the *Options Page*. This is fine if configuring one option, however multiple configurations would require the user to navigate back to the *Options Page* continuously.

File: firefoxThemes/options/options.js
Comment (Unthrottled) [reply firefoxThemes/options/options.js:33]: The `editorAccentColor` property is part of the new plugin update.

**Comment by Unthrottled** [comment]
> Since it is a small code change, there was an issue when the user selects an option in the General category

Cool! I tested it out, works as advertised.

This has been a long running PR and I have lost track of the outstanding issues, is the last thing that needs to be addressed? https://github.com/doki-theme/doki-theme-web/pull/86#discussion_r778472225

File: firefoxThemes/options/options.js
Comment (ZimCodes) [reply firefoxThemes/options/options.js:33]: The extension will take notice of the `editorAccentColor` property so long as you update the *DokiThemeDefinitions.js* file before publishing.

**Comment by ZimCodes** [comment]
@Unthrottled Yes, in accordance to *Opt-In Firefox Features*. 

Note: https://github.com/doki-theme/doki-theme-web/pull/86#discussion_r778689670

File: firefoxThemes/options/options.js
Comment (Unthrottled) [reply firefoxThemes/options/options.js]: Got it, I removed the unnecessary code then. Thanks for adding clarity!

**Comment by Unthrottled** [comment]
Alright, I think everything is squared away now.

File: firefoxThemes/options/options.js
Comment (ZimCodes) [inline firefoxThemes/options/options.js]: There's some very minor issues I found involving the secondary background button for druthers.

File: firefoxThemes/popup/popup.css
Comment (ZimCodes) [inline firefoxThemes/popup/popup.css:211]: An `onhover` effect of some sort will provide affordance to `#show-settings`.

**Review by ZimCodes — APPROVED** [review]
Although more suggestions are brought to attention, it would be better to handle them in another pr. They are not apart of the *Opt-In* feature. 

I would also like to bring up another proposal, however not sure at this point should I keep opening these as an issue or put it into a discussion board?

For now, I approve the current state of the this p. @Unthrottled Unless there's something else that needs to be added here?

**Review by ZimCodes — CHANGES_REQUESTED** [review]
My proposal has something in relation to this topic as well. What should be done with the *suggestive content* theme(s)? Should this also be considered an opt-in theme category? 

The current state of Doki Theme is not aware of this specific category and thus is blended in with the rest of the themes in all modes Doki Theme supports.

File: firefoxThemes/popup/popup.css
Comment (Unthrottled) [reply firefoxThemes/popup/popup.css]: I'll agree that it can enhance the UX a bit, but I am not a fan of the font size change & addition of word shadow. So I opted for a color change in place.

**Comment by Unthrottled** [comment]
>  What should be done with the suggestive content theme(s)? Should this also be considered an opt-in theme category?

Yeah, that should probably be added. I'll get that taken care of. I might any/all suggestive content in the future.

**Comment by Unthrottled** [comment]
Alright, since the spicy content only became available when using `mixed mode`, I just removed the content from even showing up.

File: firefoxThemes/popup/popup.css
Comment (ZimCodes) [reply firefoxThemes/popup/popup.css]: Since the popup menu contains animated elements, it is only fair the `Show Settings` link also receives an animation on hover.

Comment (ZimCodes) [reply firefoxThemes/popup/popup.css]: > So I opted for a color change in place.

For some themes, the color change is difficult to see.

https://user-images.githubusercontent.com/23222943/148705895-871d5bfd-9b92-4ad5-950f-4990870672ec.mp4

.

**Comment by ZimCodes** [comment]
Updated the documentation, removing the suggestive theme section. And also animated the link hovering effect [See this discussion](https://github.com/doki-theme/doki-theme-web/pull/86#discussion_r780847311).

File: firefoxThemes/popup/popup.css
Comment (Unthrottled) [reply firefoxThemes/popup/popup.css]: That's fair, how about just changing opacity

https://user-images.githubusercontent.com/15972415/148706777-7483086c-6aed-476f-a445-eeb226cffe58.mp4

I think I may eventually need to think of a hyperlink color for the doki theme.

**Comment by Unthrottled** [comment]
> removing the suggestive theme section.

That's fair, it's still there (for the Chrome Themes), but the user has to go out of their way to install it. So they aren't being surprised by spiciness.

**Review by ZimCodes — APPROVED** [review]

File: buildSrc/src/BuildThemes.ts
Comment (ZimCodes) [inline buildSrc/src/BuildThemes.ts:633]: There is an issue with Kanna's background not showing in the *new tab* page.

https://user-images.githubusercontent.com/23222943/148717461-9e384d5f-a8b7-4e3b-bae8-0a085596c4b7.mp4

When issue was found, I've already looked through the rest of the themes to make sure none of them have the same problem.

**Review by ZimCodes — CHANGES_REQUESTED** [review]
Found a theming issue while checking out the new hover for `Show Settings` link. Hopefully, this would be the last?

File: buildSrc/src/BuildThemes.ts
Comment (Unthrottled) [reply buildSrc/src/BuildThemes.ts:633]: Good catch! I am looking forward to refactoring the BuildThemes.ts when pulling out firefox.

**Review by ZimCodes — APPROVED** [review]

File: buildSrc/src/BuildThemes.ts
Comment (ZimCodes) [reply buildSrc/src/BuildThemes.ts:633]: Yep, this resolves it!

**Pull request merged**
